src/lib/core/CHIPTLVReader.cpp

```diff
@@ -207,35 +207,45 @@ CHIP_ERROR TLVReader::Get(uint64_t & v)
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR TLVReader::Get(double & v)
+namespace {
+template <typename T>
+CHIP_ERROR GetFloatingImpl(const TLVElementType elementType, const uint64_t elemLenOrVal, T & v)
 {
-    switch (ElementType())
+    switch (elementType)
     {
     case TLVElementType::FloatingPointNumber32: {
-        union
-        {
-            uint32_t u32;
-            float f;
-        } cvt;
-        cvt.u32 = static_cast<uint32_t>(mElemLenOrVal);
-        v       = cvt.f;
+        float f;
+        auto u32 = static_cast<uint32_t>(elemLenOrVal);
+        memcpy(&f, &u32, sizeof(f));
+        v = f;
         break;
     }
     case TLVElementType::FloatingPointNumber64: {
-        union
-        {
-            uint64_t u64;
-            double d;
-        } cvt;
-        cvt.u64 = mElemLenOrVal;
-        v       = cvt.d;
+        double d;
+        memcpy(&d, &elemLenOrVal, sizeof(d));
+        v = static_cast<T>(d);
```

Comment:
This comes up frequently when translating between loosely typed things (e.g. json values) and strictly typed things (e.g. c or c++ types).

I have generally found there are three options:

* blindly cast and potentially drop precision, underflow, or overflow (as in the above)
* permit casting, but verify that no loss of precision, underflow, or overflow has occurred (Rust style?)
* never permit casting

Only the second two options are generally safe.  But I have further found that the second option, permitting casting but throwing an error for loss of precision, underflow, or overflow, creates a testability problem for the peer because the peer can inject data that appears to show an interface to be working when in reality certain values will break.

Never permitting casting can inform test harnesses and peers that the type used could break, even if the particular value injected could in theory be losslessly cast.  This simplifies the test matrix a lot.

Reply:
Thanks for taking the time to examine this.

I presume similar concerns also extend to the existing code for handling integers, which also truncates values? Especially considering that integers also cannot saturate to ±infinity.

Reply:
Yes.  See https://github.com/project-chip/connectedhomeip/issues/5152

Reply:
I think this is more complicated than that, that not-casting a solution ends up not actually being better than then 'permit casting _safely_', which is what I think we should go down.

If you look at how TLV encodes an integer scalar for example, it auto shrink-wraps it down to the smallest width possible before transmission. That means a receiver cannot actually infer the _true_ 'originating' type of that scalar, and only sees the condensed representation.

However, a receiver in _almost_ all cases, knows the true type of the data. And as such, that's the type they'll likely pass into the TLVReader::Get() API. Should we be erroring them out because the data over the wire is modulating type dynamically? I don't think we should, because that's going to create highly variable success/failure behavior here, where in some cases, the true type matches the wire type. But when it doesn't, the call is going to error out unexpectedly.

Rather, I think we really have no choice but to permit 'safe casting'. That being said, I don't think the silent precision loss/clipping/overflow/underflow cases should be permitted. That's IMHO, egregious since it hides mistakes.

Consequently, I think we should permit casting from a 'lower width' scalar type to a 'higher width' scalar type (but not the other way). E.g you can pass a 32-bit uint in when the wire type is 16-bits. Same goes for floats/doubles.

Reply:
Let's not drag the int discussion into this.

For floats/doubles, it sounded to me like @msandstedt was proposing that we don't allow silently reading a double into a float even if the value can be represented as a float with no loss of precision.  That sounds good to me and sounds like you agree.

Silently reading a float into a double should be fine too, as you say.

Reply:
Yes, I didn't really organize my thoughts very well.  I think when I wrote the above I was thinking of a 'never cast [to narrower / less precise types]' approach (although I didn't write that).  In any case, you're right that if the wire types and destination types are not the same, this is the strictest rule that is tenable.

So really the choices are two types of 'safe casting': never permitting a cast to a narrower / less precise type or always checking that the result of a cast won't cause data to be lost.

If it's guaranteed that wire types are always the same or narrower / less precise and never wider, you don't need the more sophisticated approach that actually dynamically verifies every cast can be performed without data loss.  So I guess the question is then: is it permissible for a peer to use a double precision wire type for something where the true underlying type is single precision?  This is naive and makes the wire format bigger, not smaller.  But is it technically legal?

Reply:
The spec is not 100% clear on this, but given that they are different schema types I would say "no".  That is, if something is declared a 32-bit float in schema, or via the Matter data types, it gets encoded as a 32-bit float in the TLV (per the schema or the encoding spec respectively).

Reply:
Agreed, it shouldn't be permissible nor compatible to send a double over-the-wire when the schema prescribes a float. 

If the API caller however wants to read it into a double however, that should be permissible. That's unlikely to happen however given our code-generation scheme which I presume, faithfully follow the schema types in the generated objects and type definitions.

Reply:
Since it sounds like there is consensus to disallow reading an over-the-wire double into a float, I have updated the commit to reflect this.

Dynamically shrinking a double into a float (if there is no loss of precision) on the sender end seems possible but somewhat messy code-wise, but I don't believe that consideration affects the TLV reader logic at all.